### PR TITLE
log rmq messages to db

### PR DESCRIPTION
**Pull request by barsukovskii**

### CODE REVIEW

File: src/main/java/com/lykke/matching/engine/daos/azure/AzureMessage.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    private String msgPart3;
+    private String msgPart4;
+    private String msgPart5;
+    private String msgPart6;
+
+    public AzureMessage() {
+    }
+
+    public AzureMessage(String id, String type, String[] parts) {
+        super(generatePartitionKey(), String.format("%s_%s", id, type));
```
Comment (dmitrykrasovskih) [inline src/main/java/com/lykke/matching/engine/daos/azure/AzureMessage.java]: use Rowkey=${timestamp}_${id}, if such RowKey exists - add _001, _002, etc, check previous version of AzureMarketOrderDatabaseAccessor for example, but don't add extra _000 for first insert

File: src/main/java/com/lykke/matching/engine/daos/azure/AzureMessage.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    private static final SimpleDateFormat DATE_FORMAT = createDateFormatter();
+
+    private String msgPart1;
+    private String msgPart2;
+    private String msgPart3;
+    private String msgPart4;
+    private String msgPart5;
+    private String msgPart6;
+
```
Comment (dmitrykrasovskih) [inline src/main/java/com/lykke/matching/engine/daos/azure/AzureMessage.java]: save id and timestamp in separate fields

File: src/main/kotlin/com/lykke/matching/engine/database/azure/AzureMessageLogDatabaseAccessor.kt [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import org.apache.log4j.Logger
+
+class AzureMessageLogDatabaseAccessor(connString: String) : MessageLogDatabaseAccessor<AzureMessage> {
+
+    companion object {
+        private val LOGGER = Logger.getLogger(AzureMessageLogDatabaseAccessor::class.java.name)
+        private val METRICS_LOGGER = MetricsLogger.getLogger()
+    }
+
+    private val eventLogTable: CloudTable = getOrCreateTable(connString, "MatchingEngineMessageLog")
```
Comment (dmitrykrasovskih) [inline src/main/kotlin/com/lykke/matching/engine/database/azure/AzureMessageLogDatabaseAccessor.kt]: Pass table name as constructor param

File: src/main/kotlin/com/lykke/matching/engine/database/azure/AzureMessageLogDatabaseAccessor.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    override fun log(message: AzureMessage) {
+        try {
+            eventLogTable.execute(TableOperation.insert(message))
+        } catch (e: Exception) {
+            val errorMessage = "Unable to insert message log: ${e.message}"
+            LOGGER.error(errorMessage, e)
```
Comment (dmitrykrasovskih) [inline src/main/kotlin/com/lykke/matching/engine/database/azure/AzureMessageLogDatabaseAccessor.kt]: For error logging use Throttling logger

File: src/main/kotlin/com/lykke/matching/engine/logging/DatabaseLogger.kt [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    private fun takeAndSaveMessage() {
+        try {
+            val event = queue.take()
+            transformMessage(event)?.let {
+                dbAccessor.log(it)
+            }
+        } catch (e: Exception) {
+            val errorMessage = "Unable to write log to DB: ${e.message}"
+            LOGGER.error(errorMessage, e)
```
Comment (dmitrykrasovskih) [inline src/main/kotlin/com/lykke/matching/engine/logging/DatabaseLogger.kt]: Use throttling logger

File: src/main/kotlin/com/lykke/matching/engine/logging/MessageDatabaseLogger.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        return when (message) {
+            is BalanceUpdate -> AzureMessage(message.id, message.type, parts(message.toJson()))
+            is CashOperation -> AzureMessage(message.id, "CashOperation", parts(message.toJson()))
+            is CashSwapOperation -> AzureMessage(message.id, "CashSwapOperation", parts(message.toJson()))
+            is CashTransferOperation -> AzureMessage(message.id, "CashTransferOperation", parts(message.toJson()))
+            is LimitOrdersReport -> AzureMessage(UUID.randomUUID().toString(), "LimitOrdersReport", parts(message.toJson()))
+            is MarketOrderWithTrades ->
+                AzureMessage(message.order.id, "MarketOrderWithTrades", parts(message.toJson()))
+            else -> {
+                LOGGER.error("Unknown message type: ${this::class.java.name}")
```
Comment (dmitrykrasovskih) [inline src/main/kotlin/com/lykke/matching/engine/logging/MessageDatabaseLogger.kt]: Use throttling logger

File: src/main/kotlin/com/lykke/matching/engine/logging/MessageDatabaseLogger.kt [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+                null
+            }
+        }
+    }
+
+    private fun parts(value: String): Array<String> {
```
Comment (dmitrykrasovskih) [inline src/main/kotlin/com/lykke/matching/engine/logging/MessageDatabaseLogger.kt]: move parts method and all Azure related classes inside azure package, here should be only generic things, interfaces. If we will switch from azure to sql (or whatever) - we should change only implementation

File: src/main/kotlin/com/lykke/matching/engine/messages/MessageProcessor.kt [hunk 8]
```diff
@@ -143,15 +145,16 @@ class MessageProcessor(config: Config, queue: BlockingQueue<MessageWrapper>) : T
         val connectionsHolder = ConnectionsHolder(orderBooksQueue)
         connectionsHolder.start()
 
+        val messageLogger = MessageDatabaseLogger(AzureMessageLogDatabaseAccessor(config.me.db.messageLogConnString))
```
Comment (dmitrykrasovskih) [inline src/main/kotlin/com/lykke/matching/engine/messages/MessageProcessor.kt]: Create separate loggers for each rabbit publisher and write to separate tables

**Comment by barsukovskii** [comment]
fixed

**Pull request merged**